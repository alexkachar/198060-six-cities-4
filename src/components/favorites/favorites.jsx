import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import OfferTypes from '../../types/offer';
import Header from '../header/header.jsx';
import OfferCard from '../offer-card/offer-card.jsx';
import Footer from '../footer/footer.jsx';
import {getFavorites} from '../../store/reducers/data/selectors.js';
import DataOperation from '../../store/operations/data/data.js';
import FavoritesEmpty from '../favorites-empty/favorites-empty.jsx';
import {CARD_TYPES} from '../../helpers/constants.js';

class Favorites extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onRequestFavorites} = this.props;
    onRequestFavorites();
  }

  render() {
    const {favorites, onSetFavoriteStatus, onTitleClick} = this.props;
    const cities = [...new Set(favorites.map((offer) => offer.city.name))].sort();
    const hasFavorites = favorites.length > 0;

    return hasFavorites ? (
      <div className="page">

        <Header />

        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">

                {cities.map((city) => (
                  <li className="favorites__locations-items" key={city}>

                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>

                    <div className="favorites__places">

                      {favorites.filter((offer) => offer.city.name === city).map((offer) => (
                        <OfferCard
                          offer={offer}
                          onCardHover={() => {}}
                          onCardHoverLeave={() => {}}
                          onSetFavoriteStatus={onSetFavoriteStatus}
                          onTitleClick={onTitleClick}
                          key={offer.id}
                          cardType={CARD_TYPES.favorites}
                        />))}

                    </div>
                  </li>
                ))}

              </ul>
            </section>
          </div>
        </main>

        <Footer />
      </div>
    ) : <FavoritesEmpty />;
  }
}

const mapStateToProps = (state) => ({
  favorites: getFavorites(state),
});

const mapDispatchToProps = (dispatch) => ({
  onRequestFavorites: () => {
    dispatch(DataOperation.loadFavorites());
  }
});

export {Favorites};

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);

Favorites.propTypes = {
  favorites: PropTypes.arrayOf(OfferTypes.isRequired).isRequired,
  onSetFavoriteStatus: PropTypes.func,
  onTitleClick: PropTypes.func,
  onRequestFavorites: PropTypes.func.isRequired
};
