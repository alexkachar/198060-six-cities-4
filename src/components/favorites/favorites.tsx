import * as React from 'react';
import {connect} from 'react-redux';

import Offer from '../../interfaces/offer';
import Header from '../header/header';
import OfferCard from '../offer-card/offer-card';
import Footer from '../footer/footer';
import {getFavorites} from '../../store/reducers/data/selectors';
import {getAuthStatus} from '../../store/reducers/user/selectors';
import DataOperation from '../../store/operations/data/data';
import FavoritesEmpty from '../favorites-empty/favorites-empty';
import {AuthStatus, CardTypes} from '../../helpers/constants';
import {getCitiesList} from '../../helpers/utils';

interface Props {
  favorites: Offer[];
  authStatus: string;
  onSetFavoriteStatus: (offerId: string | number, isFavorite: boolean) => void;
  onRequestFavorites: () => void;
  onCardHover: (offer: Offer) => void;
  onCardHoverLeave: () => void;
}

class Favorites extends React.PureComponent<Props> {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onRequestFavorites} = this.props;
    onRequestFavorites();
  }

  render() {
    const {favorites, authStatus, onSetFavoriteStatus, onCardHover, onCardHoverLeave} = this.props;
    const cities = getCitiesList(favorites);
    const hasFavorites = favorites.length > 0;
    const isAuthorized = authStatus === AuthStatus.AUTH;

    return hasFavorites ? (
      <div className="page">

        <Header />

        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">

                {cities.map((city) => (
                  <li className="favorites__locations-items" key={city}>

                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>

                    <div className="favorites__places">

                      {favorites.filter((offer) => offer.city.name === city).map((offer) => (
                        <OfferCard
                          offer={offer}
                          onCardHover={onCardHover}
                          onCardHoverLeave={onCardHoverLeave}
                          onSetFavoriteStatus={onSetFavoriteStatus}
                          key={offer.id}
                          isAuthorized={isAuthorized}
                          cardType={CardTypes.FAVORITES}
                        />))}

                    </div>
                  </li>
                ))}

              </ul>
            </section>
          </div>
        </main>

        <Footer />
      </div>
    ) : <FavoritesEmpty />;
  }
}

const mapStateToProps = (state) => ({
  favorites: getFavorites(state),
  authStatus: getAuthStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  onRequestFavorites: () => {
    dispatch(DataOperation.loadFavorites());
  }
});

export {Favorites};

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
