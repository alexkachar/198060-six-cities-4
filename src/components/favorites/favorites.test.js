import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';

import {Favorites} from './favorites.jsx';
import mockStore from '../../test-data/mock-store.js';
import offersMock from '../../test-data/offers.js';

it(`Favorites renders correctly`, () => {
  const tree = renderer
    .create(

        <Provider store={mockStore}>
          <BrowserRouter>
            <Favorites
              favorites={offersMock}
              onRequestFavorites={() => offersMock} />
          </BrowserRouter>
        </Provider>, {
          createNodeMock: () => document.createElement(`div`)

        })
    .toJSON();
  expect(tree).toMatchSnapshot();
});
