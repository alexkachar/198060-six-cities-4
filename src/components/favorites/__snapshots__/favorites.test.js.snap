// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders correctly 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height={41}
              src="/img/logo.svg"
              width={81}
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  Sign In
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="place-card undefined"
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
              >
                <div
                  className="place-card__image-wrapper undefined"
                >
                  <a
                    href="#"
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height={200}
                      src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg"
                      width={260}
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €248
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height={19}
                        width={18}
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating 
                        4.4
                      </span>
                    </div>
                  </div>
                  <a
                    href="/offer/1"
                    onClick={[Function]}
                  >
                    <h2
                      className="place-card__name"
                    >
                      Penthouse, 4-5 rooms + 5 balconies
                    </h2>
                  </a>
                  <p
                    className="place-card__type"
                  >
                    hotel
                  </p>
                </div>
              </article>
              <article
                className="place-card undefined"
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
              >
                <div
                  className="place-card__image-wrapper undefined"
                >
                  <a
                    href="#"
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height={200}
                      src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg"
                      width={260}
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €248
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height={19}
                        width={18}
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating 
                        4.4
                      </span>
                    </div>
                  </div>
                  <a
                    href="/offer/2"
                    onClick={[Function]}
                  >
                    <h2
                      className="place-card__name"
                    >
                      Penthouse, 4-5 rooms + 5 balconies
                    </h2>
                  </a>
                  <p
                    className="place-card__type"
                  >
                    hotel
                  </p>
                </div>
              </article>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height={33}
        src="img/logo.svg"
        width={64}
      />
    </a>
  </footer>
</div>
`;
